cpp/pipelines/add_authority_wikidata_ids.cc

```diff
@@ -0,0 +1,195 @@
+
+std::string ltrim(const std::string &s) {
+    return std::regex_replace(s, std::regex("^\\s+"), std::string(""));
+}
+
+
+std::string rtrim(const std::string &s) {
+    return std::regex_replace(s, std::regex("\\s+$"), std::string(""));
+}
+
+
+std::string trim(const std::string &s) {
+    return ltrim(rtrim(s));
+}
+
+
+void ParseDataDnbFile(std::string input_filename, std::string output_filename) {
+    std::ifstream input_file(input_filename);
+    std::ofstream output_file(output_filename);
+    if (input_file.is_open() and output_file.is_open()) {
+        std::string line;
+        std::string act_gnd;
+        std::string act_name;
+        std::string act_wiki;
+        bool sameAs_reached(false);
+        bool read_preferred_name(false);
+        bool read_gnd_id(false);
+        while (std::getline(input_file, line)) {
+            if (line == "}, {") {
+                if (not act_wiki.empty() and not act_name.empty() and not act_wiki.empty()) {
+                    output_file << "Name: " << act_name << " GND: " << act_gnd << " Wikidata: " << act_wiki << "\n";
+                    std::cout << "Name: " << act_name << " GND: " << act_gnd << " Wikidata: " << act_wiki << "\n";
+                }
+                act_gnd = "";
+                act_name = "";
+                act_wiki = "";
+                sameAs_reached = false;
+                read_gnd_id = true;
+            } else if (read_preferred_name) {
+                read_preferred_name = false;
+                act_name = std::regex_replace(line, std::regex("(value|:|\"|@)"), "");
+                act_name = trim(act_name);
+            } else if (StringUtil::Contains(line, "info/gnd/") and read_gnd_id) {
+                read_gnd_id = false;
+                std::size_t last_slash = line.find_last_of("/");
+                act_gnd = line.substr(last_slash + 1);
+                act_gnd = std::regex_replace(act_gnd, std::regex("(\\s|,|\")"), "");
+            } else if (StringUtil::Contains(line, "www.wikidata.org/entity/") and sameAs_reached) {
+                std::size_t last_slash = line.find_last_of("/");
+                act_wiki = line.substr(last_slash + 1);
+                act_wiki = std::regex_replace(act_wiki, std::regex("(\\s|,|\")"), "");
+            } else if (StringUtil::Contains(line, "owl#sameAs")) {
+                sameAs_reached = true;
+            } else if (StringUtil::Contains(line, "preferredNameForThePerson")) {
+                read_preferred_name = true;
+            }
+        }
+        input_file.close();
+        output_file.close();
+    }
+    else 
+        LOG_ERROR("input or output files could not be opened");
+}
+
+void ParseGndWikidataMappingFile(std::string filename, std::unordered_map<std::string, std::string> * const gnd_to_wikidataid) {
+    std::ifstream file(filename);
+    if (file.is_open()) {
+        std::string line;
+        std::string act_gnd;
+        std::string act_wiki;
+        while (std::getline(file, line)) {
```

Comment:
If you would write a CSV file instead of this custom text file you could simply use TextUtil::ParseCSVFileOrDie to parse this and iterate over the lines.